kv/kvserver: TestNewRangefeedForceLeaseRetry failed

kv/kvserver.TestNewRangefeedForceLeaseRetry [failed](https://tanzanite.cluster.engflow.com/invocations/default/ebe3ef14-70ba-40ea-9923-401d1dd67fde?testReportRun=25&testReportShard=2&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on release-24.1 @ [25fab96bed932edb16c5e32e830fad62ec7383b5](https://github.com/cockroachdb/cockroach/commits/25fab96bed932edb16c5e32e830fad62ec7383b5):


```
=== RUN   TestNewRangefeedForceLeaseRetry
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestNewRangefeedForceLeaseRetry4045324648
    test_log_scope.go:81: use -show-logs to present logs inline
    replica_rangefeed_test.go:1688: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/replica_rangefeed_test.go:1688
        	Error:      	Received unexpected error:
        	            	liveness record changed while incrementing epoch for liveness(nid:2 epo:1 exp:1727929763.238261911,0); actual is liveness(nid:2 epo:1 exp:1727929763.134542165,0); is the node still live?
        	            	(1) liveness record changed while incrementing epoch for liveness(nid:2 epo:1 exp:1727929763.238261911,0); actual is liveness(nid:2 epo:1 exp:1727929763.134542165,0); is the node still live?
        	            	Error types: (1) *liveness.ErrEpochCondFailed
        	Test:       	TestNewRangefeedForceLeaseRetry
    panic.go:626: -- test log scope end --
test logs left over in: outputs.zip/logTestNewRangefeedForceLeaseRetry4045324648
--- FAIL: TestNewRangefeedForceLeaseRetry (0.52s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=25</code>
 - <code>shard=2</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/replication
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestNewRangefeedForceLeaseRetry.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-42896

--- a/pkg/kv/kvserver/replica_rangefeed_test.go
+++ b/pkg/kv/kvserver/replica_rangefeed_test.go
@@ -1837,13 +1837,24 @@ func TestNewRangefeedForceLeaseRetry(t *testing.T) {
 	// Expire the lease. Given that the Raft leadership is on n2, only n2 will be
 	// eligible to acquire a new lease.
 	log.Infof(ctx, "test expiring lease")
-	nl := n2.NodeLiveness().(*liveness.NodeLiveness)
-	resumeHeartbeats := nl.PauseAllHeartbeatsForTest()
-	n2Liveness, ok := nl.Self()
+	nl2 := n2.NodeLiveness().(*liveness.NodeLiveness)
+	resumeHeartbeats := nl2.PauseAllHeartbeatsForTest()
+	n2Liveness, ok := nl2.Self()
 	require.True(t, ok)
 	manualClock.Increment(n2Liveness.Expiration.ToTimestamp().Add(1, 0).WallTime - manualClock.UnixNano())
 	atomic.StoreInt64(&rejectExtraneousRequests, 1)
-	// Ask another node to increment n2's liveness record.
+
+	// Ask another node to increment n2's liveness record, but first, wait until
+	// n1's liveness state is the same as n2's. Otherwise, the epoch below might
+	// get rejected because of mismatching liveness records.
+	testutils.SucceedsSoon(t, func() error {
+		nl1 := n1.NodeLiveness().(*liveness.NodeLiveness)
+		n2LivenessFromN1, _ := nl1.GetLiveness(n2.NodeID())
+		if n2Liveness != n2LivenessFromN1.Liveness {
+			return errors.Errorf("waiting for node 2 liveness to converge on both nodes 1 and 2")
+		}
+		return nil
+	})
 	require.NoError(t, n1.NodeLiveness().(*liveness.NodeLiveness).IncrementEpoch(ctx, n2Liveness))
 
 	resumeHeartbeats()

